File: typescript/src/anthropic.ts
/**
 * Anthropic client wrapper that reports metadata to chaos trace.
 *
 * Construct with `new Anthropic({...})` exactly like the official SDK.
 * Adds chaostraceApiKey + callerTag args. Every call to
 * messages.create() additionally fires an async metadata report —
 * content is never sent, only timing/model/token counts.
 */
import {
  SDK_VERSION,
  newCallId,
  newRunId,
  nowIso,
  postReportAsync,
  resolveConfig,
} from './client.js'

// Anthropic SDK is a peer dependency — load it lazily so chaos trace's
// OpenAI-only users don't get an error.
function getAnthropicClass(): new (config: unknown) => unknown {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const mod = require('@anthropic-ai/sdk')
    const cls = mod.Anthropic ?? mod.default
    if (!cls) throw new Error('Anthropic class not found in @anthropic-ai/sdk')
    return cls as new (config: unknown) => unknown
  } catch {
    throw new Error(
      "chaos trace's Anthropic wrapper requires @anthropic-ai/sdk. " +
        'Install it with: npm install @anthropic-ai/sdk',
    )
  }
}

export interface AnthropicOptions {
  apiKey?: string
  baseURL?: string
  chaostraceApiKey?: string
  chaostraceApiUrl?: string
  callerTag?: string
  chaostraceRunId?: string
  [key: string]: unknown
}

export class Anthropic {
  // We return the wrapped underlying client from the constructor (a
  // legitimate JS pattern — `new Cls()` uses the returned object if
  // one is returned). The declared type is `Anthropic` to look like
  // a drop-in replacement; at runtime it's the real Anthropic instance
  // with messages.create wrapped.
  constructor(options: AnthropicOptions = {}) {
    const AnthropicClass = getAnthropicClass()
    const {
      chaostraceApiKey,
      chaostraceApiUrl,
      callerTag,
      chaostraceRunId,
      ...anthropicOptions
    } = options

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const inner = new AnthropicClass(anthropicOptions) as any
    const config = resolveConfig(chaostraceApiKey, chaostraceApiUrl, callerTag)
    const runId = chaostraceRunId ?? newRunId()
    const tag = config.callerTag ?? 'anthropic-client'

    const originalCreate = inner.messages.create.bind(inner.messages)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    inner.messages.create = async function (params: any): Promise<any> {
      const started = Date.now()
      const response = await originalCreate(params)
      const latencyMs = Date.now() - started

      try {
        const usage = (response && response.usage) || {}
        const model = params?.model || response?.model || 'unknown'
        postReportAsync(config, {
          spec_version: '1.0',
          agent_id: tag,
          run_id: runId,
          started_at: nowIso(),
          completed_at: nowIso(),
          status: 'completed',
          model,
          tool_calls: [
            {
              tool: 'anthropic.messages.create',
              outcome: 'succeeded',
              decided_at: nowIso(),
            },
          ],
          outputs: [],
          metadata: {
            framework: 'chaostrace-typescript',
            sdk_version: SDK_VERSION,
            latency_ms: latencyMs,
            input_tokens: usage.input_tokens ?? null,
            output_tokens: usage.output_tokens ?? null,
            cache_creation_input_tokens: usage.cache_creation_input_tokens ?? null,
            cache_read_input_tokens: usage.cache_read_input_tokens ?? null,
            call_id: newCallId(),
          },
        })
      } catch {
        // Telemetry failure must never affect the customer's code path.
      }

      return response
    }

    // Return the wrapped inner client. The constructor's return value
    // overrides `this` per JS spec.
    return inner as unknown as Anthropic
  }
}

File: python/chaostrace/_anthropic.py
"""Anthropic client wrapper that reports metadata to chaos trace.

Subclasses anthropic.Anthropic so all standard client APIs work
unchanged. Overrides messages.create() to capture metadata and POST
to /api/agents/report after the response returns (asynchronously, never
blocking the customer's code path).
"""
import time
from typing import Any, Optional

try:
    import anthropic as _anthropic
except ImportError as exc:
    raise ImportError(
        "chaos trace's Anthropic wrapper requires the anthropic package. "
        "Install it with: pip install anthropic"
    ) from exc

from ._client import (
    SDK_VERSION,
    _new_call_id,
    _new_run_id,
    _now_iso,
    _post_report_async,
    _resolve_config,
)


class Anthropic(_anthropic.Anthropic):
    """Drop-in replacement for anthropic.Anthropic.

    All standard Anthropic SDK functionality is preserved. Each call to
    messages.create() additionally fires an async metadata report to
    chaos trace — content is never sent, only timing, model, token
    counts, and the caller_tag you set.
    """

    def __init__(
        self,
        *args: Any,
        chaostrace_api_key: Optional[str] = None,
        chaostrace_api_url: Optional[str] = None,
        caller_tag: Optional[str] = None,
        chaostrace_run_id: Optional[str] = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(*args, **kwargs)
        self._chaostrace_config = _resolve_config(
            chaostrace_api_key, chaostrace_api_url, caller_tag
        )
        # One run_id per client lifetime by default. Customers who want
        # to group calls into discrete runs can override per-client.
        self._chaostrace_run_id = chaostrace_run_id or _new_run_id()
        self._wrap_messages_create()

    def _wrap_messages_create(self) -> None:
        original_create = self.messages.create
        config = self._chaostrace_config
        run_id = self._chaostrace_run_id
        tag = config.caller_tag or "anthropic-client"

        def wrapped_create(*args: Any, **kwargs: Any) -> Any:
            started = time.monotonic()
            response = original_create(*args, **kwargs)
            elapsed_ms = int((time.monotonic() - started) * 1000)

            try:
                usage = getattr(response, "usage", None)
                model = kwargs.get("model") or getattr(response, "model", None) or "unknown"
                payload = {
                    "spec_version": "1.0",
                    "agent_id": tag,
                    "run_id": run_id,
                    "started_at": _now_iso(),
                    "completed_at": _now_iso(),
                    "status": "completed",
                    "model": model,
                    "tool_calls": [
                        {
                            "tool": "anthropic.messages.create",
                            "outcome": "succeeded",
                            "decided_at": _now_iso(),
                        }
                    ],
                    "outputs": [],
                    "metadata": {
                        "framework": "chaostrace-python",
                        "sdk_version": SDK_VERSION,
                        "latency_ms": elapsed_ms,
                        "input_tokens": getattr(usage, "input_tokens", None) if usage else None,
                        "output_tokens": getattr(usage, "output_tokens", None) if usage else None,
                        "cache_creation_input_tokens": getattr(usage, "cache_creation_input_tokens", None) if usage else None,
                        "cache_read_input_tokens": getattr(usage, "cache_read_input_tokens", None) if usage else None,
                        "call_id": _new_call_id(),
                    },
                }
                _post_report_async(config, payload)
            except Exception:
                # Telemetry failure must never affect the customer's code path.
                pass

            return response

        self.messages.create = wrapped_create  # type: ignore[method-assign]
